### PR TITLE
Add configuration management interfaces and services

### LINKED ISSUE #1528: Support configuration management on the backend.
### What would you like to be added?

Add configuration management backend services

### Why is this needed?

Enables users to configure plug-ins on the backend.

### PR DESCRIPTION
**Pull request by hanbingleixue**
**What type of PR is this?**

Feature.

**What this PR does / why we need it?**

1、Add configuration management interfaces and services
2、Add UTs for  configuration management  interfaces and services.
3、Add the encryption and decryption tool class.
4、Add the interface for determining whether the client is connected.

**Which issue(s) this PR fixes？**

Fixes #1528 
**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [x] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java [hunk 1]
```diff
@@ -56,4 +56,14 @@ public static boolean isValidGroupName(String group) {
     public static String reBuildGroup(String group) {
         return group.replace('=', ':').replace('&', '_').replace('/', '.');
     }
+
+    /**
+     * conversion the group name to nacos group
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java]: convert

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java [hunk 2]
```diff
@@ -56,4 +56,14 @@ public static boolean isValidGroupName(String group) {
...
+     *
+     * @param group group name
+     * @return valid group
+     */
+    public static String conversionGroup(String group) {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java]: convertGroup

File: sermant-backend/src/main/java/io/sermant/backend/util/AesUtil.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (C) 2022-2022 Huawei Technologies Co., Ltd. All rights reserved.
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/util/AesUtil.java]: update

File: sermant-backend/src/test/java/io/sermant/backend/service/ConfigServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
+package io.sermant.backend.service;
```
Comment (lilai23) [inline sermant-backend/src/test/java/io/sermant/backend/service/ConfigServiceTest.java:17]: check copyright

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java [hunk 5]
```diff
@@ -141,35 +141,64 @@ public Map<String, List<String>> getConfigList(String key, String group, boolean
...
+        if (index == 0) {
+            parentPath = path.substring(0, index + 1);
+        } else {
+            parentPath = path.substring(0, index);
+        }
+        String currentPath = path.substring(index + 1);
+        if (!ifNodeExist(parentPath)) {
+            return configList;
+        }
+        List<String> strings = this.zkClient.getChildren(parentPath, false);
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: Use meaningful names.

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java [hunk 6]
```diff
@@ -141,35 +141,64 @@ public Map<String, List<String>> getConfigList(String key, String group, boolean
...
+                return accurateGetConfigListByGroupAndKey(key, path, configList);
             }
+            return fuzzyGetConfigListByGroupAndKey(key, path, configList);
         } catch (KeeperException | InterruptedException e) {
             LOGGER.error("Exception in querying configuration list", e);
         }
         return configList;
     }
 
+    private Map<String, List<String>> fuzzyGetConfigListByGroupAndKey(String key, String path, Map<String,
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: remove configList param

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java [hunk 7]
```diff
@@ -141,35 +141,64 @@ public Map<String, List<String>> getConfigList(String key, String group, boolean
...
+                    .collect(Collectors.toList());
+            if (matchSubChild.isEmpty()) {
+                continue;
+            }
+            configList.put(childPath.substring(1), matchSubChild);
+        }
+        return configList;
+    }
+
+    private Map<String, List<String>> accurateGetConfigListByGroupAndKey(String key, String path, Map<String,
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: remove configList param，and the methods should be divided into accurateGetConfigListByGroupAndKey and accurateGetConfigListByGroup

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+/**
+ * Configure Service
+ *
+ * @author zhp
+ * @since 2024-05-16
+ */
+@Service
+public class ConfigService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(HeartBeatInfoController.class);
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: LoggerFactory.getLogger(HeartBeatInfoController.class)?

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    private static final int PERIOD = 10;
+
+    @Resource
+    private DynamicConfig dynamicConfig;
+
+    private ConfigClient configClient;
+
+    private Watcher watcher;
+
+    private static boolean filterConfiguration(ConfigInfo request, ConfigInfo configInfo) {
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: configinfo is not clear

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    /**
+     * Add configuration
+     *
+     * @param configInfo Configuration information
+     * @return The result of adding configuration
+     */
+    @PostMapping("/config")
+    public Result<Boolean> addConfig(@RequestBody(required = false) ConfigInfo configInfo) {
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java]: true

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+    public Result<List<ConfigInfo>> getConfigList(ConfigInfo request, PluginType pluginType) {
+        Result<?> result = checkConnection(request);
+        if (!result.isSuccess()) {
+            return new Result<>(result.getCode(), result.getMessage());
+        }
+        ConfigClient client = getConfigClient(request.getNamespace());
+        Map<String, List<String>> configMap = client.getConfigList(request.getKey(), request.getGroup(), false);
+        List<ConfigInfo> configInfoList = new ArrayList<>();
+        for (Map.Entry<String, List<String>> entry : configMap.entrySet()) {
+            String group = entry.getKey();
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java:108]: Verify the group and key separately

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+     * @return The result of deleting configuration information
+     */
+    public Result<Boolean> deleteConfig(ConfigInfo request) {
+        Result<Boolean> checkResult = checkConnection(request);
+        if (!checkResult.isSuccess()) {
+            return checkResult;
+        }
+        String group = request.getGroup();
+        if (configClient instanceof NacosClient) {
+            group = NacosUtils.reBuildGroup(request.getGroup());
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: rebuildGroup

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+        }
+    }
+
+    /**
+     * Get Configuration Center Client
+     *
+     * @param namespace Configuration namespace
+     * @return Configuration Center Client
+     */
+    private ConfigClient getConfigClient(String namespace) {
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java:247]: getNacosConfigClient

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+        if (ConfigCenterType.ZOOKEEPER.name().equals(dynamicConfig.getDynamicConfigType())) {
+            watcher = new Watcher() {
+                @Override
+                public void process(WatchedEvent event) {
+                    if (event.getState() == Watcher.Event.KeeperState.Expired) {
+                        configClient = new ZooKeeperClient(serverAddress, timeout, this);
+                    }
+                }
+            };
+            ZooKeeperClient client = new ZooKeeperClient(dynamicConfig.getServerAddress(), dynamicConfig.getTimeout(),
```
Comment (daizhenyu) [inline sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: serverAddress，and timeout，dot need get from dynamicConfig

File: sermant-backend/src/main/java/io/sermant/backend/util/AesUtil.java (hunk 3 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/util/AesUtil.java]: done

File: sermant-backend/src/test/java/io/sermant/backend/service/ConfigServiceTest.java (hunk 4 above)
Comment (hanbingleixue) [reply sermant-backend/src/test/java/io/sermant/backend/service/ConfigServiceTest.java:17]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java (hunk 1 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java (hunk 2 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/nacos/NacosUtils.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 6 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 7 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 8 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 9 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java (hunk 10 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 11 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java:108]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 12 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 13 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java:247]: done

File: sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java (hunk 14 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/service/ConfigService.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+        if (StringUtils.isEmpty(configInfo.getPluginType())) {
+            return new Result<>(ResultCodeType.MISS_PARAM.getCode(), ResultCodeType.MISS_PARAM.getMessage());
+        }
+        Optional<PluginType> optionalPluginType = PluginType.getPluginType(configInfo.getPluginType());
+        if (!optionalPluginType.isPresent()) {
+            return new Result<>(ResultCodeType.FAIL.getCode(), "Invalid plugin name.");
+        }
+        PluginType pluginType = optionalPluginType.get();
+        if (pluginType == PluginType.OTHER) {
+            if (StringUtils.isEmpty(configInfo.getGroup())) {
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java:73]: merge these two conditional statements

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    /**
+     * Add configuration
+     *
+     * @param configInfo Configuration information
+     * @return The result of adding configuration
+     */
+    @PostMapping("/config")
+    public Result<Boolean> addConfig(@RequestBody() ConfigInfo configInfo) {
```
Comment (lilai23) [inline sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java]: `()` after `@RequestBody` can be removed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/dynamicconfig/zookeeper/ZooKeeperClient.java]: done

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java (hunk 15 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java:73]: When pluginType == PluginType.OTHER, logic processing is added and no combination is required.

File: sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java (hunk 16 above)
Comment (hanbingleixue) [reply sermant-backend/src/main/java/io/sermant/backend/controller/ConfigController.java]: done

**Review by daizhenyu — DISMISSED** [review]

**Review by lilai23 — APPROVED** [review]

**Review by daizhenyu — APPROVED** [review]

**Pull request merged**